maturity-level-two/src/Services/CoditoRepository.cs

```diff
@@ -0,0 +1,106 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+using Codit.LevelTwo.Entities;
+using Codit.LevelTwo.Extensions;
+using Microsoft.EntityFrameworkCore;
+
+
+
+namespace Codit.LevelTwo.Services
+{
+    public class CoditoRepository : ICoditoRepository
+    {
+
+        public CoditoContext _coditoContext;
+
+        public CoditoRepository(CoditoContext coditoContext)
+        {
+            _coditoContext = coditoContext;
+        }
+
+        public async Task<bool> CarExistsAsync(int id)
+        {
+            return await _coditoContext.Cars.AnyAsync(car => car.Id == id);
+        }
+
+        public async Task<IEnumerable<Car>> GetCarsAsync(CarBodyType? bodyType)
+        {
+            if (bodyType == null)
+            {
+                return await _coditoContext.Cars.OrderBy(car => car.Id).ToListAsync();
+            }
+            return await _coditoContext.Cars.Where(car => car.BodyType == bodyType).OrderBy(car => car.Id).ToListAsync();
+
+        }
+
+        public async Task<Car> GetCarAsync(int id, bool includeCustomization)
+        {
+            if (includeCustomization)
+            {
+                return await _coditoContext.Cars.Include(car => car.Customizations)
+                    .Where(car => car.Id == id).FirstOrDefaultAsync();
+            }
+
+            return await _coditoContext.Cars.Where(car => car.Id == id).FirstOrDefaultAsync();
+        }
+
+        public async Task<IEnumerable<Customization>> GetAllCustomizationsAsync()
+        {
+
+           return await _coditoContext.Customizations.OrderByDescending(cust => cust.NumberSold).ToListAsync();
+
+        }
+
+        public async Task<Customization> GetCustomizationAsync(int id)
+        {
+            return await _coditoContext.Customizations.AsNoTracking().Where(cust => cust.Id == id).FirstOrDefaultAsync();
+        }
+
+        public async Task CreateCustomizationAsync(Customization customization)
+        {
+
+             _coditoContext.Customizations. AddAsync(customization);
+            await _coditoContext.SaveChangesAsync();
+            
+        }
+
+        public async Task ApplyPatchAsync<TEntity, TDto>(TEntity entityToUpdate, TDto dto) where TEntity : class
```

Comment:
The `TDto` represents a model that is exposed by the API.  I wouldn't use that as a parameter in the repository.
What I would do, is pass in the `TEntity` in where the properties are already modified.

Reply:
issue created